src/node/internal/internal_http_client.ts

```diff
@@ -0,0 +1,441 @@
+import { OutgoingMessage } from 'node-internal:internal_http_outgoing';
+import { Agent, globalAgent } from 'node-internal:internal_http_agent';
+import type { IncomingMessageCallback } from 'node-internal:internal_http_util';
+
+const INVALID_PATH_REGEX = /[^\u0021-\u00ff]/;
+
+function validateHost(host: unknown, name: string): string {
+  if (host !== null && host !== undefined && typeof host !== 'string') {
+    throw new ERR_INVALID_ARG_TYPE(
+      `options.${name}`,
+      ['string', 'undefined', 'null'],
+      host
+    );
+  }
+  return host as string;
+}
+
+export class ClientRequest extends OutgoingMessage {
+  #abortController = new AbortController();
+  #body: Buffer[] = [];
+  #incomingMessage?: IncomingMessage;
+  #timer: number | null = null;
+
+  public _ended: boolean = false;
+
+  public timeout?: number;
+  public method: string = 'GET';
+  public path: string;
+  public host: string;
+  public protocol: string = 'http:';
+  public port: string = '80';
+  public joinDuplicateHeaders: boolean | undefined;
+  public agent: Agent | undefined;
+
+  public constructor(
+    input: string | URL | RequestOptions | null,
+    options?: RequestOptions | IncomingMessageCallback,
+    cb?: IncomingMessageCallback
+  ) {
+    super();
+
+    if (typeof input === 'string') {
+      input = urlToHttpOptions(new URL(input));
+    } else if (isURL(input)) {
+      // url.URL instance
+      input = urlToHttpOptions(input);
+    } else {
+      cb = options as IncomingMessageCallback;
+      options = input as RequestOptions;
+      input = null;
+    }
+
+    if (typeof options === 'function') {
+      cb = options;
+      options = input ?? {};
+    } else {
+      options = Object.assign(input ?? {}, options);
+    }
+
+    if (options.path) {
+      const path = String(options.path);
+      if (INVALID_PATH_REGEX.test(path)) {
+        throw new ERR_UNESCAPED_CHARACTERS('Request path');
+      }
+    }
+
+    type AgentLike = Agent | boolean | null | undefined;
+    let agent = options.agent as unknown as AgentLike;
+    // TODO(soon): Rather than using RequestOptions use our own type that includes our own Agent class type.
+    const defaultAgent =
+      (options._defaultAgent as unknown as AgentLike) || globalAgent;
+    if (agent === false) {
+      // @ts-expect-error TS2351 This expression is not constructable.
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment,@typescript-eslint/no-unsafe-call
+      agent = new defaultAgent.constructor();
+    } else if (agent == null) {
+      if (typeof options.createConnection !== 'function') {
+        agent = defaultAgent as Agent;
+      }
+    } else if (
+      typeof agent === 'object' &&
+      typeof agent.addRequest !== 'function'
+    ) {
+      throw new ERR_INVALID_ARG_TYPE(
+        'options.agent',
+        ['Agent-like Object', 'undefined', 'false'],
+        agent
+      );
+    }
+    this.agent = agent as unknown as Agent | undefined;
```

Comment:
Should that be `|| undefined`?

As is this will cause `this.agent` to be equal to 0.